tracing-opentelemetry-jdbc/src/main/java/io/micronaut/tracing/opentelemetry/instrument/jdbc/JdbcTelemetryFactory.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.tracing.opentelemetry.instrument.jdbc;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.event.BeanCreatedEventListener;
+import io.micronaut.core.annotation.Internal;
+import io.opentelemetry.api.OpenTelemetry;
+import io.micronaut.jdbc.DataSourceResolver;
+import io.opentelemetry.instrumentation.jdbc.datasource.JdbcTelemetry;
+import jakarta.inject.Singleton;
+
+import javax.sql.DataSource;
+
+/**
+ * Jdbc telemetry factory.
+ */
+@Factory
+@Internal
+final class JdbcTelemetryFactory {
+
+    /**
+     * Wraps the DataSource so OTEL can gather data for spans.
+     *
+     * @param telemetry instance of {@link OpenTelemetry}
+     * @param resolver the {@link DataSourceResolver}
+     * @return even listener for @{@link DataSource} creation
+     */
+    @Singleton
+    BeanCreatedEventListener<DataSource> otel(OpenTelemetry telemetry, DataSourceResolver resolver) {
+        return event -> {
+            DataSource dataSource = event.getBean();
+            return JdbcTelemetry.create(telemetry).wrap(resolver.resolve(dataSource));
```

Comment:
I don't think this will work as it will unwrap the datasource and break any injection points that use https://github.com/micronaut-projects/micronaut-data/blob/40b2035609740757375d56a06d48ebea76fffb70/data-connection-jdbc/src/main/java/io/micronaut/data/connection/jdbc/advice/ContextualAwareDataSource.java#L55

What has to be done is this has to declare `@Ordered` to ensure that this listener runs before `ContextualAwareDataSource`

In addition you will likely have to add a `DataSourceResolver` implementation that looks like:

```java
@Singleton
@Replaces(DelegatingDataSourceResolver.class)
class OpenTelemetryResolver extends DelegatingDataSourceResolver {
    @Override
    public DataSource resolve(DataSource dataSource) {
        if (dataSource instance OpenTelemetryDataSource otel) {
              return otel.unwrap(DataSource.class);
        } 
        return super.resolve(dataSource);
    }
}
```